### PR TITLE
AVX-595 Backport OpenDroneId from Ardupilot Copter 4.0

### PR DESCRIPTION
**Pull request by rabad-matternet**
- cherry picked OpenDroneId Copter 4.0 commits
- updated serial I/O for MAVLink communication for OpenDroneId messaging

### CODE REVIEW

File: libraries/AP_HAL_ChibiOS/hwdef/MttrCubeOrange/defaults.parm [hunk 1]
```diff
@@ -189,7 +193,7 @@ SERIAL3_PROTOCOL	5
 SERIAL4_BAUD	115
 SERIAL4_PROTOCOL	100
 SERIAL5_BAUD	57
-SERIAL5_PROTOCOL -1
+SERIAL5_PROTOCOL 2
```
Comment (viet-nguyen-matternet) [inline libraries/AP_HAL_ChibiOS/hwdef/MttrCubeOrange/defaults.parm:196]: For my understanding. What does 2 mean in this instance?

**Review by viet-nguyen-matternet — APPROVED** [review]
looks great, you mentioned we have no method of testing this with current hardware. Does that mean testing between Cube -> BLE module will be halted until we get a revised AP board? That could be a blocker in development.

File: libraries/AP_HAL_ChibiOS/hwdef/MttrCubeOrange/defaults.parm (hunk 1 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/hwdef/MttrCubeOrange/defaults.parm:196]: it means MAVLink2

**Comment by tridge** [comment]
> looks great, you mentioned we have no method of testing this with current hardware.

I have tested it with Robbie on my board.
for local testing you need a Cube carrier board without ADSB. Just use a carrier from a CubeBlack

File: libraries/AP_OpenDroneID/AP_OpenDroneID_DroneCAN.cpp [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+#define ODID_COPY(name) msg.name = pkt.name
+#define ODID_COPY_STR(name) do { for (uint8_t i = 0; i<sizeof(pkt.name) && pkt.name[i]; i++) msg.name.push_back(pkt.name[i]); } while(0)
+
+
+void AP_OpenDroneID::dronecan_send_location(AP_UAVCAN *uavcan)
+{
+    dronecan::remoteid::Location msg {};
+    const auto &pkt = pkt_location;
+    ODID_COPY_STR(id_or_mac);
+    ODID_COPY(status);
```
Comment (viet-nguyen-matternet) [inline libraries/AP_OpenDroneID/AP_OpenDroneID_DroneCAN.cpp:183]: @rabad-matternet @tridge , just checking, will AP set this field automatically in the event of emergency status?

Where Emergency Status is:
UAS unable to recover from an uncontrolled descent
UAS unable to recover from loss of control of the flight trajectory

Is this triggered by an FTS? Can we modify this field manually for testing?

File: libraries/AP_HAL_ChibiOS/hwdef/MttrCubeOrange/defaults.parm (hunk 1 above)
Comment (rabad-matternet) [reply libraries/AP_HAL_ChibiOS/hwdef/MttrCubeOrange/defaults.parm:196]: 1 = MAVLink 1
2 = MAVLink 2
5 = GPS

File: libraries/AP_UAVCAN/dsdl/dronecan/remoteid/20033.System.uavcan [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+#
+uint8[<=20] id_or_mac
+
+uint8 ODID_OPERATOR_LOCATION_TYPE_TAKEOFF = 0 # The location of the operator is the same as the take-off location
+uint8 ODID_OPERATOR_LOCATION_TYPE_LIVE_GNSS = 1 # The location of the operator is based on live GNSS data
+uint8 ODID_OPERATOR_LOCATION_TYPE_FIXED = 2 # The location of the operator is a fixed location
+uint8 operator_location_type
+
+uint8 ODID_CLASSIFICATION_TYPE_EU = 0
+uint8 classification_type
```
Comment (viet-nguyen-matternet) [inline libraries/AP_UAVCAN/dsdl/dronecan/remoteid/20033.System.uavcan:13]: @rabad-matternet @tridge this classification type field makes it seem like there is only the Classification Type EU. However in the F3411-22a block it mentions:

Classification Type: Bits [4..2]
0 = Undeclared
1 = European Union
2–7 = Reserved

Is this implementation incorrect?

**Pull request merged**